### PR TITLE
disable password confirmation with SSO

### PR DESCRIPTION
**Pull request by schiessle**
Disable password confirmation in case of single-sign-on.
In case of SSO password confirmation doesn't work because Nextcloud can't check the users password.

### CODE REVIEW

**Comment by blizzz** [comment]
Can we have unit tests for the  middle ware that test positive as well as negative cases?

**Comment by rullzer** [comment]
Pfff right. I knew I disliked all this functionality in 1 middleware. I'd like to just add an additional one that we can properly test.

**Comment by rullzer** [comment]
* rebased
* Moved password confirmation to own middleware
* Added new middleware tests

**Review by rullzer — APPROVED** [review]

File: lib/private/AppFramework/Middleware/Security/PasswordConfirmationMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		if ($this->reflector->hasAnnotation('PasswordConfirmationRequired')) {
+			$user = $this->userSession->getUser();
+			$backendClassName = '';
+			if ($user !== null) {
+				$backendClassName = $user->getBackendClassName();
+			}
+
+			$lastConfirm = (int) $this->session->get('last-password-confirm');
+			// we can't check the password against a SAML backend, so skip password confirmation in this case
+			if ($backendClassName !== 'user_saml' && $lastConfirm < ($this->timeFactory->getTime() - (30 * 60 + 15))) { // allow 15 seconds delay
```
Comment (blizzz) [inline lib/private/AppFramework/Middleware/Security/PasswordConfirmationMiddleware.php]: I don't like it is hard-coded… if we can test for an instance (→ OCP) that would be much better. Perhaps for 14, for now (and backporting?) good enough.

**Review by blizzz — APPROVED** [review]
cannot test, but code looks good and i am confident with the unit tests

**Pull request merged**